Introduce local methods in SharedData for async map, counter and lock

pk-work opened the pull request:
Closes #2654

In the current implementation of SharedData it is not possible to get a
local lock, counter or async map if a cluster manager is used.
Fortunately all the necessary logic is already in place, so that this
change is minimal invasive.

tsegismont commented:
@pk-work thanks for the PR. This is a good start. To complete it we need:

- shared data documentation update
- test (like two clustered nodes may acquire a local lock with same name at the same time)

I would also prefer not to break the feature into pieces (local locks in 3.7 and the rest in 4.0). Can you also contribute the changes for local counters and local async map?

pk-work commented:
@tsegismont  thanks for your feedback.

Of course I can try to contribute the changes for local counters and local async map. It's just a matter of time, at the moment I'm working on a do or die project and the timeline is very tough. So I can't spent that much time in the next 3 months.

But I will check the effort for this and will tell you if I think I can do it for 3.7 or not. If not, I will contribute it in 4.0.

pk-work commented:
Hi @tsegismont,

I have now updated the documentation, but I don't know how to check the generated HTML and how it will really look like. I also added getLocalCounter and getLocalAsyncMap.

I have not yet added tests, I will do this the next days. Can you already check, if the rest is fine for you?

tsegismont commented on src/main/asciidoc/shareddata.adoc:
```diff
@@ -38,6 +38,9 @@ Here's an example of using a shared local map:
 Asynchronous shared maps allow data to be put in the map and retrieved locally when Vert.x is not clustered.
 When clustered, data can be put from any node and retrieved from the same node or any other node.
 
+In case that Vert.x is clustered, but a local asynchronous map is needed, you can call
```
I would prefer to move this below the `IMPORTANT` callout and say that if you don't need the AsyncMap to be shared between nodes when Vert.x is not clustered, you can use `getLocalAsyncMap`.

tsegismont commented on src/main/asciidoc/shareddata.adoc:
```diff
@@ -92,6 +95,8 @@ is obtained.
 
 To obtain a lock use {@link io.vertx.core.shareddata.SharedData#getLock(java.lang.String,io.vertx.core.Handler)}.
 
+To obtain a local lock, even if Vert.x is clustered, use {@link io.vertx.core.shareddata.SharedData#getLocalLock(java.lang.String,io.vertx.core.Handler)}.
```
Can you rework this part of the doc to look like the map one? An `IMPORTANT` callout indicating that in clustered mode, the lock relies on distributed data structures ...etc and if you don't need this, invoke `getLocalLock` ?

tsegismont commented on src/main/asciidoc/shareddata.adoc:
```diff
@@ -129,4 +134,8 @@ You obtain an instance with {@link io.vertx.core.shareddata.SharedData#getCounte
 Once you have an instance you can retrieve the current count, atomically increment it, decrement and add a value to
 it using the various methods.
 
+In case that Vert.x is clustered, {@link io.vertx.core.shareddata.SharedData#getCounter(java.lang.String,io.vertx.core.Handler)} will return
```
Same comment as for locks.

pk-work commented:
I updated now the documentation. I will push the missing tests asap.

pk-work commented:
@tsegismont  Hi I found now time to add test cases. Can you check, if this change is fine for you? Then I would also create a PR for 3.7 branch. Then this could maybe also go into 3.7.

tsegismont commented:
@pk-work thank you. I just reviewed the PR, it's going to 3.7. I did some changes in the doc and will merge in a few minutes.

tsegismont commented:
Merged in 9496285cc

pk-work commented:
@tsegismont Thank you so much :) This is my first PR for Vert.x which is really merged now :) Feels very good.

Will you create a PR for 3.7?

tsegismont commented:
Keep them coming! ;)

No I cherry-picked the changes over to 3.7 branch already
